File: CivitasJ/src/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author ramonjtorres
 */
public class CivitasJuego {
    
    private int indiceJugadorActual;
    GestorEstados gestorEstados;
    EstadosJuego estado;
    Tablero tablero;
    MazoSorpresas mazo;
    ArrayList<Jugador> jugadores;
    
    public CivitasJuego(ArrayList<String> nombres){
    
    jugadores = new ArrayList<Jugador>();
    gestorEstados = new GestorEstados();
    for (String s:nombres){
        Jugador j = new Jugador(s);
        jugadores.add(j);
    }    
    
    estado = gestorEstados.estadoInicial();
    indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
    mazo = new MazoSorpresas();
    this.inicializarMazoSorpresas(tablero);
    this.inicializarTablero(mazo);  
    

    
    }
    
    private void avanzaJugador(){}
    
    private void actualizarInfo(){
    
        System.out.println("Estado: "+this.estado);
        System.out.println("Información Jugador: "+this.infoJugadorTexto());
    
    }
    
    public boolean cancelarHipoteca(int ip){
    
        return jugadores.get(this.indiceJugadorActual).cancelarHipoteca(ip);
    }
    
    public boolean comprar(){ return false;}
    
    public boolean construirCasa(int ip){
        
        return jugadores.get(this.indiceJugadorActual).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
    
        return jugadores.get(this.indiceJugadorActual).construirHotel(ip);
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
    
        while(tablero.getPorSalida() > 0){
        
            jugadorActual.pasaPorSalida();
        }
    }
    
    public boolean finalDelJuego(){
        
        for(Jugador j:jugadores){
        
            if(j.enBancarrota()){
            
                return true;
            }
        }
        
        return false;
        
    }
    
    public Casilla getCasillaActual(){
        int i = this.jugadores.get(indiceJugadorActual).getNumCasillaActual();
        return this.tablero.getCasilla(i);
    }
    
    
    public Jugador getJugadorActual(){
        
        return jugadores.get(this.indiceJugadorActual);
    }
    
    public boolean hipotecar(int ip){
        
        return jugadores.get(this.indiceJugadorActual).hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        return jugadores.get(this.indiceJugadorActual).toString();
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        this.mazo = new MazoSorpresas();
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL,tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA,4,tablero));//a la carcel
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA,5,tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA,10,tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL,mazo));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR,tablero,-50,"El jugador debe pagar a cada uno de los demas jugadores 50€"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR,tablero,50,"Cada jugador te debe pagar 50€"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL,tablero,30,"Recibes 30€ por cada casa y hotel en propiedad"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL,tablero,-30,"Cobras 30€ por cada casa y hotel en propiedad"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR,tablero,-100,"Pagas 100€ por gastos de limpieza"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR,tablero,100,"Has ganado un premio al hotel más limpio recibe 100€"));

        
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        
        tablero = new Tablero(4);//la carcel

        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Ronda de Valencia", 10, (float) 0.5,25,50,20)));
 
        tablero.añadeCasilla(new Casilla((float)100, "Impuesto"));
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Lavapies",10, (float) 0.5,25,50,20)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Cuatro Caminos",20, (float) 0.6,30,70,40)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Reina Victoria",20, (float) 0.6,30,70,40)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Bravo Murillo",30, (float) 0.7,35,90,60)));
        
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa"));
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Alberto Aguilera",40, (float) 0.7,35,90,80)));
        
        tablero.añadeCasilla(new Casilla("Parking"));
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Fuencarral",40, (float) 0.8,40,110,80)));
        
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa"));
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Felipe II",50, (float) 0.8,40,110,100)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Velázquez",50, (float) 0.8,45,130,100)));
        
        tablero.añadeJuez();
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Puerta del Sol",70, (float) 0.8,45,160,100)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Alcalá",70, (float) 0.8,50,160,100)));
        
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa"));
        
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Paseo del Prado",100, (float) 0.8,60,250,120)));
        
        }
    
    private void pasarTurno(){
    
        this.indiceJugadorActual = (this.indiceJugadorActual + 1) % this.jugadores.size();
    }

    private ArrayList<Jugador> ranking(){
        
        ArrayList<Jugador> ranking = new ArrayList();
        Jugador aux = new Jugador("AUXILIAR");
        ranking = this.jugadores;
        int i = 1;
        
        while(i < ranking.size()-1){
            
            if(ranking.get(i-1).compareTo(ranking.get(i)) < 0){
               
                aux = ranking.get(i-1);
                ranking.remove(i-1);
                ranking.add(i, aux);                
           }
            
           i++;
        }
        
        return ranking;
    }
    
    public boolean salirCarcelPagando(){
        
        return jugadores.get(this.indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        
        return jugadores.get(this.indiceJugadorActual).salirCarcelTirando();        
    }
    
    public Operaciones_juego siguientePaso(){return null;}
    
    public void siguientePasoCompletado(Operaciones_juego operacion){
    
        this.estado = this.gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), this.estado, operacion);
    }
    
    public boolean vender(int ip){
        
        return jugadores.get(this.indiceJugadorActual).vender(ip);
    }
    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        
        String j1 = "David";
        String j2 = "Ramón";
        
        ArrayList <String> todos = new ArrayList();
        
        todos.add(j1);
        todos.add(j2);
        
        CivitasJuego cj = new CivitasJuego(todos);
        
        TituloPropiedad propiedad = new TituloPropiedad("Ronda de Valencia",10, (float) 0.5,25,50,20);
        
        cj.indiceJugadorActual = 1;
        
        cj.getJugadorActual().propiedades.add(propiedad);
        
        propiedad.comprar(cj.getJugadorActual());
        
        cj.actualizarInfo();
        
        System.out.println("Debe devolver true, al construir una casa: " + cj.getJugadorActual().propiedades.get(0).construirCasa(cj.getJugadorActual()));
        System.out.println("Debe devolver true, al construir una casa: " + cj.getJugadorActual().propiedades.get(0).construirCasa(cj.getJugadorActual()));
        System.out.println("Debe devolver true, al construir una casa: " + cj.getJugadorActual().propiedades.get(0).construirCasa(cj.getJugadorActual()));
        System.out.println("Debe devolver true, al construir una casa: " + cj.getJugadorActual().propiedades.get(0).construirCasa(cj.getJugadorActual()));
        System.out.println("Debe devolver true, al construir un hotel: " + cj.getJugadorActual().propiedades.get(0).construirHotel(cj.getJugadorActual()));
        
        System.out.println("Debe devolver informacion del jugador actual: " + cj.infoJugadorTexto());
        
        System.out.println("Debe salir 6:" + cj.tablero.nuevaPosicion(20, 6));
        System.out.println("Debe salir 6:" + cj.tablero.nuevaPosicion(20, 6));
        System.out.println("Debe salir 6:" + cj.tablero.nuevaPosicion(20, 6));
        System.out.println("Debe salir 10:" + cj.tablero.nuevaPosicion(0, 10));
        
        cj.contabilizarPasosPorSalida(cj.getJugadorActual());
        
        System.out.println("Debe dar 10350, al pasar 3 veces por salida (7350 + 3*1000): " + cj.getJugadorActual().getSaldo());
        
        System.out.println("Debe dar que estamos en la casilla 0, Salida: " + cj.getCasillaActual());
        
        System.out.println("Debe dar Ramon: " + cj.getJugadorActual());
        
        cj.pasarTurno();
        
        System.out.println("Debe dar David: " + cj.getJugadorActual());
        
        System.out.println("Debe dar en primera posicion Ramón, en segunda David: " + cj.ranking());
        
        cj.getJugadorActual().modificarSaldo(20000);
        
        cj.pasarTurno();
        
        System.out.println("Debe dar en primera posicion David, en segunda Ramón: " + cj.ranking());
        
        cj.siguientePasoCompletado(Operaciones_juego.AVANZAR);
        
        System.out.println("Debe dar DESPUES_AVANZAR: " + cj.estado);
        
        cj.siguientePasoCompletado(Operaciones_juego.COMPRAR);
        
        System.out.println("Debe dar DESPUES_COMPRAR: " + cj.estado);
        
        System.out.println("Debe dar true al vender: " + cj.vender(0));
        
        System.out.println("Debe dar false, no hay fin del juego: " + cj.finalDelJuego());
        
        cj.getJugadorActual().modificarSaldo(-20000);
        
        System.out.println("Debe dar true, hay fin del juego: " + cj.finalDelJuego());

    }
}

File: CivitasR/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "Tipo_Sorpresas"

module Civitas
  class Civitas_Juego
    
    @mazo = Mazo_Sorpresas.new
    @tablero = Tablero.new(20)
    @jugadores = Array.new()
    
    public
    def initialize(nombres)
      var = 0
      while (var < nombres.length())
        j = Jugador.new(nombres.at(var))
        @jugadores.push(j)
        var = var + 1
      end
      
      @gestorEstados.estado_inicial
      @indice_jugador_actual = Dado.quien_empieza(jugadores.length())
      @mazo = Mazo_Sorpresas.new
      inicializar_mazo_sorpresas(@tablero)
      inicializar_tablero(@mazo)
    end
    
    private
    def avanza_jugador()
      
    end
    
    private
    def actualizar_info()
      puts "Estado: " + @estado
      puts "Información Jugador: " + info_jugador_texto()
    end
    
    public
    def cancelar_hipoteca(ip)
      @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(ip)
    end
    
    public
    def comprar()
      
    end
  
    public
    def construir_casa(ip)
      @jugadores.at(@indice_jugador_actual).construir_casa(ip)
    end
    
    public
    def construir_hotel(ip)
      @jugadores.at(@indice_jugador_actual).construir_hotel(ip)
    end
    
    private
    def contabilizar_pasos_por_salida(jugador_actual)
      while(@tablero.get_por_salida > 0)
        jugador_actual.pasa_por_salida
      end
    end
    
    public
    def final_del_juego()
      var = 0
      while(var<@jugadores.length)
        if(@jugadores.at(var).en_bancarrota)
          return true
        end
        var = var + 1
      end
      return false
    end
    
    public
    def get_casilla_actual()
      var i = @jugadores.at(@indice_jugador_actual).num_casilla_actual
      return @tablero.get_casilla(i)
    end
    
    public
    def get_jugador_actual()
      return @jugadores.at(@indice_jugador_actual)
    end
    
    public
    def hipotecar(ip)
      return @jugadores.at(@indice_jugador_actual).hipotecar(ip)
      
    end
    
    public
    def info_jugador_texto()
       @jugadores.at(@indice_jugador_actual).to_string
    end
    
    private
    def inicializar_mazo_sorpresas(tablero)
      @mazo = Mazo_Sorpresas.new
      @mazo.al_mazo(Sorpresa.new(Tipo_Sorpresas::IRCARCEL, tablero))
      @mazo.al_mazo(Sorpresa.sorpresa_tablero(Tipo_Sorpresas::IRCASILLA,4,tablero))
      @mazo.al_mazo(Sorpresa.sorpresa_tablero(Tipo_Sorpresas::IRCASILLA,5,tablero))
      @mazo.al_mazo(Sorpresa.sorpresa_tablero(Tipo_Sorpresas::IRCASILLA,10,tablero))
      @mazo.al_mazo(Sorpresa.sorpresa_mazo(Tipo_Sorpresas::SALIRCARCEL,@mazo))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresas::PORJUGADOR,tablero,-50,"El jugador debe pagar a cada uno de los demas jugadores 50€"))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresa::PORJUGADOR,tablero,50,"Cada jugador te debe pagar 50€"))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresa::PORCASAHOTEL,tablero,30,"Recibes 30€ por cada casa y hotel en propiedad"))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresa::PORCASAHOTEL,tablero,-30,"Cobras 30€ por cada casa y hotel en propiedad"))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresa::PAGARCOBRAR,tablero,-100,"Pagas 100€ por gastos de limpieza"))
      @mazo.al_mazo(Sorpresa.sorpresa_valor(Tipo_Sorpresa::PAGARCOBRAR,tablero,100,"Has ganado un premio al hotel más limpio recibe 100€"))

    end
    
    private
    def inicializar_tablero(mazo)
        
        @tablero = Tablero.new(4);#la carcel

        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Ronda de Valencia",10, 0.5,25,50,20)))
 
        @tablero.añade_casilla(Casilla.casilla_cantidad(0, "Impuesto"));
        
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Lavapies",10,0.5,25,50,20)));
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Cuatro Caminos",20,0.6,30,70,40)));
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Reina Victoria",20,0.6,30,70,40)));
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Bravo Murillo",30,0.7,35,90,60)));
        
        @tablero.añade_casilla(Casilla.casilla_mazo(mazo, "Sorpresa"));
        
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Alberto Aguilera",40,0.7,35,90,80)));
        
        @tablero.añade_casilla(Casilla.new("Parking"));
        
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Fuencarral",40,0.8,40,110,80)));
        
        @tablero.añade_casilla(Casilla.casilla_mazo(mazo, "Sorpresa"));
        
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Felipe II",50,0.8,40,110,100)));
        @tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Velázquez",50,0.8,45,130,100)));
        
        @tablero.añade_juez;
        
        tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Puerta del Sol",70,0.8,45,160,100)));
        tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Alcalá",70,0.8,50,160,100)));
        
        tablero.añade_casilla(Casilla.casilla_mazo(mazo, "Sorpresa"));
        
        tablero.añade_casilla(Casilla.casilla_titulo(Titulo_Propiedad.new("Paseo del Prado",100,0.8,60,250,120)));
        
    end
    
    private
    def pasar_turno()
      @indice_jugador_actual = (@indice_jugador_actual + 1) % @jugadores.length
    end
    
    private
    def ranking()
      ranking = Array.new
      aux = Jugador.new("aux")
      ranking = @jugadores
      i = 0
      j = 1
      while(j<ranking.length)
        if(0>ranking.at(i).compare_to(ranking.at(j)))
          aux = ranking.at(i)
          ranking.insert(i, ranking.at(j))
          ranking.delete_at(j)
          ranking.insert(j,aux)
          ranking.delete_at(j+1)
          i = i + 1
          j = j + 1
        end
      end
      
      return ranking
    end
    
    public
    def salir_carcel_pagando()
      return @jugadores.at(@indice_jugador_actual).salir_carcel_pagando
    end
    
    public
    def salir_carcel_tirando()
      return @jugadores.at(@indice_jugador_actual).salir_carcel_tirando
    end
    
    public
    def siguiente_paso()
      
    end
    
    public
    def siguiente_paso_completado(operacion)
      @estado = @gestor_estados.siguiente_estado(@jugadores.at(@indice_jugador_actual), @estado, operacion)
    end
    
    public
    def vender(ip)
      return @jugadores.at(@indice_jugador_actual).vender(ip)
    end
    
  end
end
